qgis/examples/elevation_tile_downloader/core/task.py

```diff
@@ -1,117 +1,122 @@
-                       Qgis,
-                       QgsRasterLayer,
-                       QgsProject,
-                       QgsNetworkAccessManager,
-                       QgsNetworkReplyContent,
-                       QgsGeometry
-                       )
-from PyQt5.QtCore import QFileInfo, QUrl, pyqtSignal
-from qgis.PyQt.QtNetwork import QNetworkRequest
-from typing import List
+from typing import List, Optional, Set
 from pathlib import Path
-import traceback
 import gzip
 from io import BytesIO
 
+from qgis.gui import QgisInterface
+from qgis.core import (
+    QgsTask,
+    QgsMessageLog,
+    Qgis,
+    QgsRasterLayer,
+    QgsProject,
+    QgsNetworkAccessManager,
+    QgsNetworkReplyContent,
+    QgsGeometry,
+)
+from qgis.utils import iface
+from qgis.PyQt.QtCore import QUrl
+from qgis.PyQt.QtWidgets import QMessageBox
+from qgis.PyQt.QtNetwork import QNetworkRequest
+
+iface: QgisInterface
 
-class DownloadTask(QgsTask):
-    next_tile = pyqtSignal(int)
 
-    def __init__(self, grid: List[QgsGeometry], output_dir, add=False):
+class DownloadTask(QgsTask):
+    def __init__(self, grid: Set[QgsGeometry], output_dir: Path):
         super().__init__("Tile download task", QgsTask.CanCancel)
         self.grid = grid
         self.output_dir = output_dir
-        self.exception = None
-        self.traceback = None
-        self.total = len(self.grid)
-        self.add = add
+
+        self.exception: Optional[Exception] = None
         self.tiles: List[Path] = []
-        self.nam = QgsNetworkAccessManager()
 
     def run(self):
-        QgsMessageLog.logMessage(
-            message='Started task {}'.format(self.description()),
-            level=Qgis.Info
-        )
+        """Do the work."""
+        nam = QgsNetworkAccessManager()
+
+        # Don't raise any error during processing, relay error reporting to self.finished()
         try:
-            for i, poly in enumerate(self.grid):
-                self.next_tile.emit(i)
-                x, y = poly.boundingBox().xMinimum(), poly.boundingBox().yMinimum()
-                self.prepare_dir(y, self.output_dir)
-                tile_name = '%s%02d%s%03d.hgt' % ('S' if y < 0 else 'N', abs(y), 'W' if x < 0 else 'E', abs(x))
-                dir_name = '%s%02d' % ('S' if y < 0 else 'N', abs(y))
-                url = QUrl(f"http://s3.amazonaws.com/elevation-tiles-prod/skadi/{dir_name}/{tile_name}.gz")
-                request_ = QNetworkRequest(url)
-                dest_directory = Path(self.output_dir, dir_name)
-                filepath = Path(dest_directory, tile_name)
-                if not filepath.is_file():
-                    reply: QgsNetworkReplyContent = self.nam.blockingGet(request_)
-                    length = int(reply.rawHeader(bytes('Content-Length'.encode())))
-                    stream = BytesIO(reply.content())
-                    with gzip.GzipFile(fileobj=stream, mode='rb') as gz:
-                        with open(filepath, 'wb') as f:
```

Comment:
ah another point here: you're getting the content length after the download actually happened. `blockingGet` does just that: block until it got all the data. so the `while` further down just runs through the entire response immediately, since all downloaded data is already in memory. not sure how that fits with your observation that it goes to 90% really fast and then hangs a little (I'd have expected the other way around actually). there's also the QNetworkAccessManager->get() method which we could've used actually. I think I never tried it as I didn't know the Qgis equivalent inherits from the Qt one. what it does is asynchronous requesting, so really it could've been a replacement for QgsTask all along. just another way of doing it I guess, I like our way now more than real asynchronous requesting. which is also not thread-safe, the way you did it here is.